src/Symfony/Component/Utf8/Utf8Trait.php

```diff
@@ -0,0 +1,354 @@
+    {
+        if ('' === $needle) {
+            return;
+        }
+
+        if (false === $part = mb_strrichr($this->string, $needle, $beforeNeedle, 'UTF-8')) {
+            return;
+        }
+
+        $result = clone $this;
+        $result->string = $part;
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reverse()
+    {
+        $result = clone $this;
+        $result->string = implode('', array_reverse(iterator_to_array($this)));
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replace($from, $to, int &$count = null)
+    {
+        if (is_string($from)) {
+            $from = array($from);
+        } elseif (!is_array($from)) {
+            throw new InvalidArgumentException('Search pattern must be a valid string or array of strings.');
+        }
+
+        if (is_string($to)) {
+            if (!preg_match('//u', $to)) {
+                throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $to));
+            }
+        } elseif (is_array($to)) {
+            foreach ($to as $replacement) {
+                if (!preg_match('//u', $replacement)) {
+                    throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $replacement));
+                }
+            }
+        } else {
+            throw new InvalidArgumentException('Pattern replacement must be a valid string or array of strings.');
+        }
+
+        foreach ($from as $i => $regex) {
+            if (!preg_match('//u', $regex)) {
+                throw new InvalidArgumentException(sprintf('Given pattern "%s" is not a valid UTF-8 encoded string.', $regex));
+            }
+        }
+
+        $result = clone $this;
+        $result->string = str_replace($from, $to, $this->string, $count);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function replaceIgnoreCase($from, $to, int &$count = null)
+    {
+        if (is_string($from)) {
+            $from = array($from);
+        } elseif (!is_array($from)) {
+            throw new InvalidArgumentException('Search pattern must be a valid string or array of strings.');
+        }
+
+        if (is_string($to)) {
+            if (!preg_match('//u', $to)) {
+                throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $to));
+            }
+        } elseif (is_array($to)) {
+            foreach ($to as $replacement) {
+                if (!preg_match('//u', $replacement)) {
+                    throw new InvalidArgumentException(sprintf('Given pattern replacement "%s" is not a valid UTF-8 encoded string.', $replacement));
+                }
+            }
+        } else {
+            throw new InvalidArgumentException('Pattern replacement must be a valid string or array of strings.');
+        }
```

Comment:
Same problem as before, second argument can be an array only if first argument is an array.